salt/metalk8s/orchestrate/downgrade/precheck.sls

```diff
@@ -0,0 +1,48 @@
+{%- set dest_version = pillar.orchestrate.dest_version %}
+
+{#- When downgrading saltenv should be the newest version #}
+{%- set nodes_versions = pillar.metalk8s.nodes.values() | map(attribute='version') | list %}
+{%- do nodes_versions.sort(cmp=salt.pkg.version_cmp, reverse=True) %}
+{%- set expected = nodes_versions | first %}
+{%- if saltenv != 'metalk8s-' ~  expected %}
+
+Invalid saltenv "{{ saltenv }}" consider using "metalk8s-{{ expected }}":
+  test.fail_without_changes
+
+{%- else %}
+
+Correct saltenv "{{ saltenv }}" for downgrade to "{{ dest_version }}":
+  test.succeed_without_changes
+
+{%- endif %}
+
+{%- set dest = (dest_version|string).split('.') %}
+{%- for node, values in pillar.metalk8s.nodes.items() %}
+
+  {#- As we don't know how many minor version we do before a new major,
+      never block downgrade between major version #}
+  {%- set current = (values['version']|string).split('.') %}
+  {%- if dest[0]|int == current[0]|int
+     and current[1]|int - dest[1]|int > 1 %}
+
+Unable to downgrade from more than 1 version, Node {{ node }} from {{ values['version'] }} to {{ dest_version }}:
+  test.fail_without_changes
+
+  {%- elif salt.pkg.version_cmp(dest_version, values['version']) == 1 %}
+
+Node {{ node }} ignored, already in {{ values['version'] }} older than {{ dest_version }}:
+  test.succeed_without_changes
+
+  {%- elif dest_version != values['version']|string %}
+
+Node {{ node }} will be downgraded from {{ values['version'] }} to {{ dest_version }}:
+  test.succeed_without_changes
+
+  {%- else %}
```

Comment:
For accuracy's sake we could use `{%- elif salt.pkg.version_cmp(dest_version, values['version']) == 1` here and a relevant failing state name in the `{%- else %}`, where we should theoretically never end up.

Also, I'm not sure I have all the pieces on expected version values but what if we have some more trailing stuff? Do we treat `2.1.1-pw` like we do `2.1.1`? (see our current 2.0.0**-dev**)

Reply:
Talking with @gdemonet , treating versions with labels the same as versions without is fine for me. Maybe we can strip labels when we compare versions, in order to actually check `0`, `-1` and `1` explicitly?

Reply:
> For accuracy's sake we could use `{%- elif salt.pkg.version_cmp(dest_version, values['version']) == 1` here and a relevant failing state name in the `{%- else %}`, where we should theoretically never end up.

Here it's more `{%- elif salt.pkg.version_cmp(dest_version, values['version']) == 0 %}`

> Maybe we can strip labels when we compare versions, in order to actually check `0`, `-1` and `1` explicitly?

I don't get what do you mean ? Btw with `pkg.cmp_version` it's the same
```
[root@single-node ~]# salt-call --local pkg.version_cmp 2.1.1-pw 2.1.1
local:
    0
```
But something weird
```
[root@single-node ~]# salt-call --local pkg.version_cmp 2.1.1-pw 2.1.1-dev
local:
    1
```

But for me `2.1.1` should be superior to `2.1.1-whatever`, as `2.1.1-*` (rc, dev, pw, ... and `2.1.1` should be a released version).

Reply:
Done